golang/cosmos/x/vibc/types/receiver.go

```diff
@@ -0,0 +1,177 @@
+package types
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/Agoric/agoric-sdk/golang/cosmos/vm"
+	channeltypes "github.com/cosmos/ibc-go/v6/modules/core/04-channel/types"
+
+	"github.com/cosmos/ibc-go/v6/modules/core/exported"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+var (
+	_ vm.PortHandler           = Receiver{}
+	_ exported.Acknowledgement = (*rawAcknowledgement)(nil)
+)
+
+type ReceiverImpl interface {
+	ReceiveSendPacket(ctx sdk.Context, packet exported.PacketI) (uint64, error)
+	ReceiveWriteAcknowledgement(ctx sdk.Context, packet exported.PacketI, ack exported.Acknowledgement) error
+	ReceiveChanOpenInit(ctx sdk.Context, order channeltypes.Order, hops []string, sourcePort, destinationPort, version string) error
+	ReceiveWriteOpenTryChannel(ctx sdk.Context, packet exported.PacketI, order channeltypes.Order, connectionHops []string, version string) error
+	ReceiveChanCloseInit(ctx sdk.Context, sourcePort, sourceChannel string) error
+	ReceiveBindPort(ctx sdk.Context, sourcePort string) error
+	ReceiveTimeoutExecuted(ctx sdk.Context, packet exported.PacketI) error
+}
+
+type Receiver struct {
+	impl ReceiverImpl
+}
+
+func NewReceiver(impl ReceiverImpl) Receiver {
+	return Receiver{
+		impl: impl,
+	}
+}
+
+type portMessage struct { // comes from swingset's IBC handler
+	Type              string              `json:"type"` // IBC_METHOD
+	Method            string              `json:"method"`
+	Packet            channeltypes.Packet `json:"packet"`
+	RelativeTimeoutNs uint64              `json:"relativeTimeoutNs,string"`
+	Order             string              `json:"order"`
+	Hops              []string            `json:"hops"`
+	Version           string              `json:"version"`
+	Ack               []byte              `json:"ack"`
+}
+
+func stringToOrder(order string) channeltypes.Order {
+	switch order {
+	case "ORDERED":
+		return channeltypes.ORDERED
+	case "UNORDERED":
+		return channeltypes.UNORDERED
+	default:
+		return channeltypes.NONE
+	}
+}
+
+func orderToString(order channeltypes.Order) string {
+	switch order {
+	case channeltypes.ORDERED:
+		return "ORDERED"
+	case channeltypes.UNORDERED:
+		return "UNORDERED"
+	default:
+		return "NONE"
+	}
+}
+
+type rawAcknowledgement struct {
+	data []byte
+}
+
+func (r rawAcknowledgement) Acknowledgement() []byte {
+	return r.data
+}
+
+func (r rawAcknowledgement) Success() bool {
+	return true
+}
+
+func (ir Receiver) Receive(cctx context.Context, str string) (ret string, err error) {
+	// fmt.Println("ibc.go downcall", str)
+	ctx := sdk.UnwrapSDKContext(cctx)
+	impl := ir.impl
+
+	msg := new(portMessage)
+	err = json.Unmarshal([]byte(str), &msg)
+	if err != nil {
+		return ret, err
```

Comment:
I _know_ `ret` is really initialized to `""`, but it _feels_ uninitialized. `:-)`
```suggestion
		return "", err
```